### PR TITLE
Implement menu part of the Lookup component

### PR DESCRIPTION
**Pull request by guergana**
Bug: T265008

### CODE REVIEW

File: tokens/properties/components/LookupMenu.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+				"value": "{transition.interaction.menu.duration.value}"
+			},
+			"transition-property" : {
+				"value": "{transition.interaction.menu.property.value}"
+			},
+			"transition-timing-function" : {
+				"value": "{transition.timing-function.ease.value}"
+			}
+		},
+		"label" : {
```
Comment (jakobw) [inline tokens/properties/components/LookupMenu.json:53]: Shouldn't this be nested inside "item"? It's the label of a menu item, not the label of the menu.

File: tokens/properties/components/LookupMenu.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+				"value": "{font.weight.bold.value}"
+			},
+			"line-height": {
+				"value": "{font.line-height.style.label.value}"
+			},
+			"font-color": {
+				"value": "{font.color.base.value}"
+			}
+		},
+		"description" : {
```
Comment (jakobw) [inline tokens/properties/components/LookupMenu.json:70]: same as above

File: vue-components/src/components/Lookup.vue [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+import ValidationMessage from './ValidationMessage.vue';
+import Input from './Input.vue';
+import LookupMenu from './LookupMenu.vue';
+
+/**
+ * Text input fields are form elements that let users input and edit values in the form of text.
```
Comment (jakobw) [inline vue-components/src/components/Lookup.vue:28]: Copy-pasta from `TextInput`. Needs its own description.

File: vue-components/src/components/Lookup.vue [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+	data() {
+		return {
+			focused: false,
+			inputValue: '',
+		};
+	},
+	props: {
+		menuItems: {
+			type: Array,
+			default: (): [] => [],
```
Comment (jakobw) [inline vue-components/src/components/Lookup.vue:43]: `[]` literally means empty array, which is technically correct for the default?! I'd have expected this to eventually declare it to be a list of menu items, though.

File: tokens/properties/components/LookupMenu.json (hunk 1 above)
Comment (guergana) [reply tokens/properties/components/LookupMenu.json:53]: yes

File: tokens/properties/components/LookupMenu.json (hunk 2 above)
Comment (guergana) [reply tokens/properties/components/LookupMenu.json:70]: agree

File: vue-components/src/components/Lookup.vue (hunk 3 above)
Comment (guergana) [reply vue-components/src/components/Lookup.vue:28]: worked on

File: vue-components/src/components/Lookup.vue (hunk 4 above)
Comment (guergana) [reply vue-components/src/components/Lookup.vue:43]: we discussed it, we think it is fine for now.

File: vue-components/src/components/Lookup.vue [hunk 5]
```diff
@@ -0,0 +1,123 @@
+<template>
+	<label :class="[ 'wikit', 'wikit-Lookup', `wikit-Lookup--${width}` ]">
+		<div class="wikit-Lookup__label">{{ label }}</div>
+		<Input
+			@input.native="inputValue = $event.target.value"
+			@focus.native="focused = true"
```
Comment (itamargiv) [inline vue-components/src/components/Lookup.vue:6]: I might be missing something here, but is it really necessary to reimplement the native focus and blur with a controlled `focus` prop? I don't see it being used elsewhere within the component behaviour.

File: vue-components/src/components/LookupMenu.vue [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	props: {
+		menuItems: {
+			type: Array,
+			default: (): [] => [],
+		},
+	},
+	watch: {
+		menuItems(): void {
+			const menuItems = this.$el.querySelectorAll( '.wikit-LookupMenu__item' );
+			if ( menuItems.length > 6 ) {
```
Comment (itamargiv) [inline vue-components/src/components/LookupMenu.vue:31]: I see why 6 but it could be considered a magic number, is it possible to add a comment explaining it here (that the menu should be scrollable with more than 6 items long, I assume)?

File: vue-components/src/components/LookupMenu.vue [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	},
+} );
+</script>
+
+<style lang="scss">
+$base: '.wikit-LookupMenu';
+
+#{$base} {
+	min-width: 100%;
+	width: 200%;
```
Comment (itamargiv) [inline vue-components/src/components/LookupMenu.vue:45]: I am a bit confused by this, as these values are not encoded in design tokens, though I'm not sure if they should be?

**Review by itamargiv — CHANGES_REQUESTED** [review]
Looks great, thank you. Requesting changes because of the `label` `div` confusion, ther rest are just questions - comments. 

Also did we decide not to add browser tests to the components of this hike? Is it something we are reserving for later? as I see none were added for any of the components of this hike.

**Comment by micgro42** [comment]
> Also did we decide not to add browser tests to the components of this hike? Is it something we are reserving for later? as I see none were added for any of the components of this hike.

There should definitely be browser tests. Though as I understand it, it was decided to implement them when the full intended functionality of the component is available. That is tracked in an extra sub-ticket: https://phabricator.wikimedia.org/T265014

File: vue-components/src/components/LookupMenu.vue (hunk 7 above)
Comment (SaiSan-WMDE) [reply vue-components/src/components/LookupMenu.vue:52]: I had the same doubt while specifying the component. I guess that using alias tokens to define the width of the menu would be consistent, and convenient in case we make this a reusable element (comparable to the input).

File: tokens/properties/components/LookupMenu.json [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+			},
+			"color" : {
+				"value": "{border.color.base.default.value}"
+			}
+		},
+		"box-shadow": {
+			"value": "{box-shadow.elevation.menu.value}"
+		},
+		"item" : {
+			"padding-vertical": {
```
Comment (micgro42) [inline tokens/properties/components/LookupMenu.json:24]: This should probably be `spaceing-vertical` or something similar

Comment (guergana) [reply tokens/properties/components/LookupMenu.json:24]: they are referenced as padding-vertical and padding-horizontal in the design files. don't know what to make of this.

Comment (micgro42) [reply tokens/properties/components/LookupMenu.json:24]: No, I don't think that is has to be the same as the base token, either.
Rather that it should not needlessly preclude the specific implementation, and apparently there is some work going on removing some implementation detail from earlier token names. OTOH, if that term (i.e. `padding`) is a good way to talk about it from a design perspective, then it's probably OK.

File: vue-components/src/components/Lookup.vue [hunk 9]
```diff
@@ -0,0 +1,123 @@
+<template>
+	<label :class="[ 'wikit', 'wikit-Lookup', `wikit-Lookup--${width}` ]">
+		<div class="wikit-Lookup__label">{{ label }}</div>
```
[the remark this answers is not in the document]
Comment (guergana) [reply vue-components/src/components/Lookup.vue:3]: yes. we need to do something similar to this here: https://github.com/wmde/wikit/pull/228/files

[the remark this answers is not in the document]
Comment (jakobw) [reply vue-components/src/components/Lookup.vue]: Agree! Rebased on top of #228 and done!

**Comment by jakobw** [comment]
Added some slightly nicer to look at examples for the story.

I think this is ready for a first look, @SaiSan-WMDE! Note that this doesn't contain any of the expected behavior beyond expanding and collapsing the Lookup menu when interacting with it. Keyboard navigation, changing the text in the input field upon selecting, and filtering will come later.

File: vue-components/src/components/LookupMenu.vue [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	props: {
+		menuItems: {
+			type: Array,
+			default: (): [] => [],
+		},
+	},
+	watch: {
+		menuItems(): void {
+			const menuItems = this.$el.querySelectorAll( '.wikit-LookupMenu__item' );
+			// the height automatically adjusts for up to 6 elements, then shows a scrollbar
```
Comment (itamargiv) [inline vue-components/src/components/LookupMenu.vue:38]: Shouldn't the comment say 7 (0 based index :P )?

**Review by itamargiv — COMMENTED** [review]
Looking good, thanks for the changes, one small comment

File: vue-components/src/components/LookupMenu.vue (hunk 10 above)
Comment (jakobw) [reply vue-components/src/components/LookupMenu.vue:38]: It's correct, I think!?

It checks for `if ( menuItems.length > maxNumberOfElementsDisplayed )` i.e. it does nothing special (= grows to the content height) for up to 6, and if there are more then it cuts off at `menuItems[ maxNumberOfElementsDisplayed ].offsetTop` which is the top of the 7th element. So there are 6 displayed, right?

Comment (guergana) [reply vue-components/src/components/LookupMenu.vue:38]: I also think it is correct with 6.

Comment (itamargiv) [reply vue-components/src/components/LookupMenu.vue:38]: looking at the storybook I see seven list items.

Comment (jakobw) [reply vue-components/src/components/LookupMenu.vue:39]: Ah, you're right! Fixed, thanks :+1:

File: tokens/properties/components/LookupMenu.json (hunk 8 above)
Comment (guergana) [reply tokens/properties/components/LookupMenu.json:24]: I asked @SaiSan-WMDE and we thought that since this is a conceptual issue, we can probably leave it like this and discuss this topic with the design team and change it retroactively when a consensus is achieved.

**Review by itamargiv — APPROVED** [review]
\o/ Is done! Thank you!

**Comment by guergana** [comment]
> \o/ Is done! Thank you!
> 
> 

I must be dreaming!

**Review by jakobw — APPROVED** [review]
Rebased and squashed the commits with the latest update. LGTM :rocket:
